rfcs/0145-doc-strings.md

```diff
@@ -0,0 +1,499 @@
+- (-) Is visually less distinctive.
+- (-) Indentation is more complex and visually less present.
+    - (+) It would be most intuitive if its indentation logic implements Nix's magic multiline strings.
+- (-) Takes up more vertical space
+- (+) Takes up less horizontal space. (In contrast to, e.g., `##` lines do not have to be prefixed)
+- (+) Only one character to change the semantics of the whole comment. 
+    - (+) Allows to add and remove things from the documentation quickly.
+    - (-) Accidentally adding/removing could happen.
+    
+</details>
+
+
+## CommonMark as the content of doc-comments
+
+**Observing**: The use of CommonMark, a widely recognized and standardized format for documents, is prevalent in the documentation of code and software libraries.
+
+**Considering**: Doc-comments' content should be intuitive to read and write and straightforward to render. Furthermore it should follow established conventions in the nix ecosystem.
+
+**Decision**: CommonMark is the content of all doc-comments.
+
+> Markdown is the most accepted and understood format for writing documentation in tech projects. Also, following the existing RFC-72 is highly profitable and consistent for the Nix ecosystem.
+
+<details>
+<summary>Arguments</summary>
+
+- (+) CommonMark is the official format for Nix documentation; Decided in [RFC72](https://github.com/NixOS/rfcs/blob/master/rfcs/0072-commonmark-docs.md).
+    - (+) It Would be consistent if this RFC builds upon the previous one.
+    - (+) Further Arguments for CommonMark, in general, can be found in RFC72
+- (+) Allows copy-paste from and to markdown files. We allow easy refactoring if documentation arises and needs to be split into separate files.
+- (-) Strictly binding doc comments content to commonMark might restrict users.
+    - (+) Users/Tools can still use regular comments or develop alternative solutions.
+- (-) CommonMark does not specify the current rich features from nixdoc, such as predefined sections and structures that could be used as a source for automated toolings. Such as types for type-checking or examples to run automated-tests
+    - (+) Future tools can still build their conventions on top of this RFC. They might not directly specify them in an RFC but be a tool developers choose for a specific codebase. However, we have yet to get that tools. So it is good when rich features remain unspecified.
+  
+</details>
+
+## Single-line doc-comments (do not exist)
+
+**Observing**: Nix offers two variants of comments; single- (`#`) and multi-line comments (`/* */`). There may be use cases where it is desirable to have a form of single-line comments subtyped for doc-comment purposes.
+
+**Considering**: Single-line comment for documentation. (Starting a doc-comment with `##`)
+
+**Decision**: Single-line comments (starting with `##`) **cannot be used** in any form for documentation puposes.
```

Comment:
Kinda sad that this option won't be available.

Doxygen has `//<`, Haskell has `-- ^`, and they are really useful.

Reply:
The main reason we decide to omit `##` was because of the visual conflict with markdowns `# Heading`
Its sad yes, but i think its okay. The benefit is, that there is only one doc-comment syntax which will benefit the consistency.

We still understand the benefits (as mentioned in the section) single line doc comments would have.

Reply:
I don't think it would be ambiguous if you have spacing, like
```nix
## Some paragraph
## 
## # Some title
## Some paragraph
```

It could be a requirement for single-line doc comments to start with `## ` (including the space)

Reply:
Okay I'll bring up the topic to the final shepherd meeting. 
@amjoseph-nixpkgs @infinisil  whats your personal opinion about this? If we define '## ' with whitespace this also means '####' which are used in a lot of places for visual seperation are not doc-comments which i think is desireable.

Reply:
I think the RFC should at least be changed to not forbid single-line doc comments, but I'd be totally on-board with also defining that this is how single-line doc comments need to look like.

I do think this is a somewhat significant addition to the RFC, I'd prefer another FCP when the shepherd team is done discussing this.

Reply:
> The main reason we wanted to omit `##`

No problem; just use some other sigil instead.

I don't have any opinion on *what* sequence of characters ought to be used to indicate "this comment documents the thing before it, not the thing after it".  Just that some sequence exists.

Reply:
> This comment documents the thing before it, not the thing after it

Oh (understood you wrong) You mean the possibility of placing doc-comments **after** the `documentable-node`.

Like:

```nix
{
   a = 1; #< The number one
   b = x: y: x * y; #< The product function
}
```

We didn't specify that and we very likely will not add it to the RFC.

I did consider this in a very early stage of this rfc but after talking with some people we came up with the decision to not include that. it is still possible to document everything, which is the goal of the RFC.

If you think its necessary, we can provide an extra "decision" section, why we did not specify it.

Reply:
> placing doc-comments after

FWIW, this is somehow hard to implement this. Because there must be a complex rule to specify which node the comment belongs to, and also cannot be implemented by single-pass parser actions.

For example, for the code showed above by @hsjobeki 

We see the comment `#< The number one` should be attached to `1`, before the comment. But why cannot it attach to `b`? I only see trivia tokens between them. You may say: hmm, there is a newline `\n`, okay, so this token becomes a different from other whitespaces.

Reply:
> visual conflict with markdowns

Maybe we can choose `#!` to start single-line doc-comments ?